In [23]:
import pandas as pd
import numpy as np
from matplotlib import pyplot as plt

In [27]:
df = pd.read_csv('/content/nova_pay_combined.csv')
df.head()

,transaction_id,customer_id,timestamp,home_country,source_currency,dest_currency,channel,amount_src,amount_usd,fee,...,ip_risk_score,kyc_tier,account_age_days,device_trust_score,chargeback_history_count,risk_score_internal,txn_velocity_1h,txn_velocity_24h,corridor_risk,is_fraud
0,fee8542d-8ee6-4b0d-9671-c294dd08ed26,402cccc9-28de-45b3-9af7-cc5302aa1f93,2022-10-03 18:40:59.468549+00:00,US,USD,CAD,ATM,278.19,278.19,4.25,...,0.123,standard,263,0.522,0,0.223,0,0,0.0,0
1,bfdb9fc1-27fe-4a85-b043-4d813d679259,67c2c6b3-ef0a-4777-a3f1-c84a851bb6ad,2022-10-03 20:39:38.468549+00:00,CA,CAD,MXN,web,208.51,154.29,4.24,...,0.569,standard,947,0.475,0,0.268,0,1,0.0,0
2,fc855034-3ea5-4993-9afa-b511d93fe5e8,6d0d9b27-fa26-45f8-93b1-2df29d182d9c,2022-10-03 23:02:43.468549+00:00,US,USD,CNY,mobile,160.33,160.33,2.70,...,0.437,enhanced,367,0.939,0,0.176,0,0,0.0,0
3,2cf8c08e-42ec-444d-a755-34b9a2a0a4ca,7bd5200c-5d19-44f0-9afe-8b339a05366b,2022-10-04 01:08:53.468549+00:00,US,USD,EUR,mobile,59.41,59.41,2.22,...,0.594,standard,147,0.551,0,0.391,0,0,0.0,0
4,d907a74d-b426-438d-97eb-dbe911aca91c,70a93d26-8e3a-4179-900c-a4a7a74d08e5,2022-10-04 09:35:03.468549+00:00,US,USD,INR,mobile,200.96,200.96,3.61,...,0.121,enhanced,257,0.894,0,0.257,0,0,0.0,0


In [28]:
df.shape

(11400, 26)

In [29]:
df.columns

Index(['transaction_id', 'customer_id', 'timestamp', 'home_country',
       'source_currency', 'dest_currency', 'channel', 'amount_src',
       'amount_usd', 'fee', 'exchange_rate_src_to_dest', 'device_id',
       'new_device', 'ip_address', 'ip_country', 'location_mismatch',
       'ip_risk_score', 'kyc_tier', 'account_age_days', 'device_trust_score',
       'chargeback_history_count', 'risk_score_internal', 'txn_velocity_1h',
       'txn_velocity_24h', 'corridor_risk', 'is_fraud'],
      dtype='object')

DATA DICTIONARY FOR NOVA PAY TRANSACTION DATASET

This is an explanation of each field in the dataset.





| Column Name | Data Type | Description | Example |
|-------------|-----------|-------------|---------|
| transaction_id | String | Unique identifier for each transaction | TXN102345 |
| customer_id | String | Unique identifier for the customer initiating the transaction | CUST5678 |
| timestamp | Datetime | Date and time when the transaction occurred | 2025-11-01 10:23:45 |
| home_country | String | Country where the customer account is registered | UK |
| source_currency | String | Currency used by the sender | GBP |
| dest_currency | String | Currency received by the recipient | USD |
| channel | Categorical | Platform used for the transaction (mobile, web, API) | mobile |
| amount_src | Float | Transaction amount in the source currency | 500.00 |
| amount_usd | Float | Transaction amount converted to USD | 620.50 |
| fee | Float | Transaction fee charged by the platform | 5.50 |
| exchange_rate_src_to_dest | Float | Exchange rate applied from source to destination currency | 1.24 |
| device_id | String | Unique identifier for the device used | DEV98765 |
| new_device | Binary | Indicates if the device is new (1 = Yes, 0 = No) | 1 |
| ip_address | String | IP address used for the transaction | 192.168.0.1 |
| ip_country | String | Country detected from the IP address | Germany |
| location_mismatch | Binary | Indicates if IP country differs from home country | 1 |
| ip_risk_score | Float | Risk score assigned to the IP address | 0.72 |
| kyc_tier | Integer | Customer KYC verification level | 2 |
| account_age_days | Integer | Number of days since account creation | 365 |
| device_trust_score | Float | Trust score assigned to the device | 0.85 |
| chargeback_history_count | Integer | Number of previous chargebacks | 1 |
| risk_score_internal | Float | Internal fraud risk score | 0.67 |
| txn_velocity_1h | Integer | Number of transactions in the past 1 hour | 3 |
| txn_velocity_24h | Integer | Number of transactions in the past 24 hours | 12 |
| corridor_risk | Float | Risk score of the country transaction corridor | 0.55 |
| is_fraud | Binary | Fraud label (1 = Fraud, 0 = Legitimate) | 0 |

In [30]:
#checking for missing data
df.isnull().sum()

,0
transaction_id,0
customer_id,0
timestamp,29
home_country,0
source_currency,0
dest_currency,0
channel,0
amount_src,0
amount_usd,305
fee,295


Checking the missing values in each variable, we can confirm that:

- timestamp has 29 missing values

- amount_usd has 305 missing values

- fee has 295 missing values

- ip_address has 305 missing values

- ip_country has 301 missing values

- kyc_tier has 300 missing values

- device_trust_score has 295 missing values

In [31]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 11400 entries, 0 to 11399
Data columns (total 26 columns):
 #   Column                     Non-Null Count  Dtype  
---  ------                     --------------  -----  
 0   transaction_id             11400 non-null  object 
 1   customer_id                11400 non-null  object 
 2   timestamp                  11371 non-null  object 
 3   home_country               11400 non-null  object 
 4   source_currency            11400 non-null  object 
 5   dest_currency              11400 non-null  object 
 6   channel                    11400 non-null  object 
 7   amount_src                 11400 non-null  object 
 8   amount_usd                 11095 non-null  float64
 9   fee                        11105 non-null  float64
 10  exchange_rate_src_to_dest  11400 non-null  float64
 11  device_id                  11400 non-null  object 
 12  new_device                 11400 non-null  bool   
 13  ip_address                 11095 non-null  obj

From the above inspection we can see that timestamp is stored as a wrong data type, an object data type instead of datetime format. Also amount_src variable is also stored as object data type rather than float or integer.

In [33]:
#checking for label imbalance
df['is_fraud'].value_counts(normalize=True).reset_index()

,is_fraud,proportion
0,0,0.912544
1,1,0.087456


The is_fraud variable which happens to be our target variable is unevenly distributed. About 91% of the target variable confirms transaction is legitimate while 9% confirms that transaction is_fraud.

**SUMMARY/OBSERVATIONS**

- The composition of the dataset is 11400 rows and 26 columns

- All fieldnames follows the standard variable naming conventions

- An initial inspection of the dataset revealed inconsistencies in the data types of certain variables. Specifically, the timestamp variable is currently stored as an object data type instead of the appropriate datetime format, which is necessary for accurate time-based analysis. Similarly, the amount_src variable is also stored as an object data type, whereas it should ideally be represented as a numeric format (integer or float) to allow for proper numerical computations and analysis. These inconsistencies will need to be corrected during the data preprocessing stage.

- An assessment of missing values across the dataset revealed that several variables contain incomplete records. Specifically, the timestamp variable contains 29 missing values, while amount_usd and ip_address each contain 305 missing values. Additionally, the fee and device_trust_score variables both have 295 missing values, ip_country has 301 missing values, and kyc_tier contains 300 missing values. These missing values will require appropriate data cleaning and preprocessing to ensure the reliability and accuracy of subsequent analysis and modeling.

- The target variable is_fraud is highly imbalanced, with approximately 91% of transactions classified as legitimate and only 9% identified as fraudulent. This imbalance introduces a risk of model bias toward the majority class, where machine learning algorithms may prioritize predicting legitimate transactions while failing to detect fraudulent ones. As a result, relying solely on accuracy could produce misleading results. To mitigate this issue, techniques such as resampling methods and appropriate evaluation metrics should be used to ensure the model effectively identifies fraudulent transactions.